### PR TITLE
Add `prefer-async-await` rule (fixes #123)

### LINKED ISSUE #123: `prefer-async-await` rule
I see a lot of people doing:

```js
test(t => {
  return fetch().then(data => {
    t.is(data, 'foo');
  });
});
```

When this would be much succinct:

```js
test(async t => {
  t.is(await fetch(), 'foo');
});
```

Would be useful to have a rule to enforce using async/await.

### PR DESCRIPTION
**Pull request by jfmengels**
Add `prefer-async-await` rule (fixes #123)

### CODE REVIEW

File: docs/rules/prefer-async-await.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Prefer using async/await instead of returning a Promise
+
+Translations: [Français](https://github.com/avajs/ava-docs/blob/master/fr_FR/related/eslint-plugin-ava/docs/rules/prefer-async-await.md)
+
+AVA comes with built-in support for async functions (async/await). This allows you to write shorter and clearer tests.
```
Comment (sindresorhus) [inline docs/rules/prefer-async-await.md]: Can you linkify `async functions` to http://www.2ality.com/2016/02/async-functions.html

File: docs/rules/prefer-async-await.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+## Fail
+
+```js
+import test from 'ava';
+
+test(t => {
+	return foo().then(res => {
+			t.is(res, 1);
+		});
```
Comment (sindresorhus) [inline docs/rules/prefer-async-await.md]: Indentation looks wonky

File: rules/prefer-async-await.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		return false;
+	}
+
+	const callee = node.callee;
+	if (callee.property.type === 'Identifier' &&
+		callee.property.name === 'then'
+	) {
+		return true;
+	}
+	return containsThen(callee.object);
```
Comment (sindresorhus) [inline rules/prefer-async-await.js]: Super nitpick, but add an empty line 19.

File: rules/prefer-async-await.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		ava.isInTestNode
+	])(node => {
+		if (node.body.type !== 'BlockStatement') {
+			return;
+		}
+
+		const statements = node.body.body;
+		const returnStatement = statements.find(statement => statement.type === 'ReturnStatement');
+		if (returnStatement && containsThen(returnStatement.argument)) {
+			// Now go and find a .then()...
```
Comment (sindresorhus) [inline rules/prefer-async-await.js]: I don't get the comment, is it in the wrong place?

File: test/prefer-async-await.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+ruleTester.run('prefer-async-await', rule, {
+	valid: [
+		header + 'test(t => { t.is(1, 1); });',
+		header + 'test(t => { foo(); });',
+		header + 'test(t => { return foo(); });',
+		header + 'test(t => { foo().then(fn); });',
+		header + 'test(t => { function foo() { return foo().then(fn); } });',
+		header + 'test(t => foo().then(fn));',
+		// This should be an error, needs improvement
```
Comment (sindresorhus) [inline test/prefer-async-await.js]: `This` => `this` for consistency with the below

File: rules/prefer-async-await.js (hunk 4 above)
Comment (jfmengels) [inline rules/prefer-async-await.js]: Ah no, that was me forgetting to remove a comment I left for myself.

File: test/prefer-async-await.js (hunk 5 above)
Comment (sindresorhus) [inline test/prefer-async-await.js]: Can you make this a `TODO` comment or open an issue about it so we don't forget?

Comment (jfmengels) [inline test/prefer-async-await.js]: I'll add a TODO. I want to get back to working on using ESLint's `context.getScope()` (first to get macros working), and after that, we'll have to re-check which rules could benefit from this "technique".

File: test/prefer-async-await.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+ruleTester.run('prefer-async-await', rule, {
+	valid: [
+		header + 'test(t => { t.is(1, 1); });',
+		header + 'test(t => { foo(); });',
+		header + 'test(t => { return foo(); });',
+		header + 'test(t => { foo().then(fn); });',
+		header + 'test(t => { function foo() { return foo().then(fn); } });',
+		header + 'test(t => foo().then(fn));',
+		// TODO this should be an error, needs improvement
```
Comment (sindresorhus) [inline test/prefer-async-await.js]: `// TODO` => `// TODO:`

**Comment by sindresorhus** [comment]
✨ Perfect :)

**Pull request merged**